File: py/pe/pe21.py
#!/usr/bin/env python
"""
Evaluate the sum of all amicable pairs under 10000.
"""

from utils import spd

def pe21(limit=10000):
    """
    >>> pe21()
    (31626, [(220, 284), (1184, 1210), (2620, 2924), (5020, 5564), (6232, 6368)])
    """
    ami = []
    s = 0
    for i in range(2, limit + 1):
        t = spd(i)
        if i < t and i == spd(t):
            s += i + t
            ami.append((i, t))
    return (s, ami)

if __name__ == "__main__":
    import doctest
    doctest.testmod()

    try:
        while True:
            s = input('> ')
            n = int(s)
            print(pe21(n))
    except (SyntaxError, EOFError, KeyboardInterrupt, NameError):
        pass

File: cpp/pe/src/pe21.cc
#define _CRT_SECURE_NO_WARNINGS 1
#include "../include/pe21.h"
#include "../include/spd.h"

#include <iostream>

namespace pe {
  using namespace std;

  static bool debug_on = false;

  uint32_t pe21(uint32_t n) {
    uint32_t sum = 0;
    
    for (uint32_t i = 2; i < n + 1; ++i) {
      uint32_t t = spd(i);
      if (i < t && i == spd(t)) {
        if (debug_on) {
          cout << "(" << i << ", " << t << ")" << endl;
        }
        sum += i + t;
      }
    }
    
    return sum;
  }
}